src/mlpack/tests/decision_tree_test.cpp

```diff
@@ -454,112 +719,77 @@ BOOST_AUTO_TEST_CASE(SimpleGeneralizationTest)
+  arma::Row<size_t> trainingLabels = l.subvec(0, 4999);
+  arma::Row<size_t> testLabels = l.subvec(5000, 9999);
 
-    spiralDataset(0, i) = x;
-    spiralDataset(1, i) = y;
+  // Build the tree.
+  DecisionTree<> tree(trainingData, di, trainingLabels, 5, 10);
 
-    // Set categorical features c1 and c2.
-    if (i < 2000)
-    {
-      spiralDataset(2, i) = c1[1].Random()[0];
-      spiralDataset(3, i) = c2[1].Random()[0];
-      labels[i] = 1;
-    }
-    else if (i < 4000)
-    {
-      spiralDataset(2, i) = c1[3].Random()[0];
-      spiralDataset(3, i) = c2[3].Random()[0];
-      labels[i] = 3;
-    }
-    else if (i < 6000)
-    {
-      spiralDataset(2, i) = c1[2].Random()[0];
-      spiralDataset(3, i) = c2[2].Random()[0];
-      labels[i] = 2;
-    }
-    else if (i < 8000)
-    {
-      spiralDataset(2, i) = c1[0].Random()[0];
-      spiralDataset(3, i) = c2[0].Random()[0];
-      labels[i] = 0;
-    }
-    else
-    {
-      spiralDataset(2, i) = c1[4].Random()[0];
-      spiralDataset(3, i) = c2[4].Random()[0];
-      labels[i] = 4;
-    }
-  }
+  // Now evaluate the accuracy of the tree.
+  arma::Row<size_t> predictions;
+  tree.Classify(testData, predictions);
 
-  // Now create the dataset info.
-  data::DatasetInfo di(4);
-  di.Type(2) = data::Datatype::categorical;
-  di.Type(3) = data::Datatype::categorical;
-  // Set mappings.
-  di.MapString<double>("0", 2);
-  di.MapString<double>("1", 2);
-  di.MapString<double>("2", 2);
-  di.MapString<double>("3", 2);
-  di.MapString<double>("0", 3);
-  di.MapString<double>("1", 3);
+  BOOST_REQUIRE_EQUAL(predictions.n_elem, testData.n_cols);
+  size_t correct = 0;
+  for (size_t i = 0; i < testData.n_cols; ++i)
+    if (testLabels[i] == predictions[i])
+      ++correct;
 
-  // Now shuffle the dataset.
-  arma::uvec indices = arma::shuffle(arma::linspace<arma::uvec>(0, 9999,
-      10000));
-  arma::mat d(4, 10000);
-  arma::Row<size_t> l(10000);
-  for (size_t i = 0; i < 10000; ++i)
-  {
-    d.col(i) = spiralDataset.col(indices[i]);
-    l[i] = labels[indices[i]];
-  }
+  // Make sure we got at least 70% accuracy.
+  const double correctPct = double(correct) / double(testData.n_cols);
+  BOOST_REQUIRE_GT(correctPct, 0.70);
+}
 
+/**
+ * Test that we can build a decision tree with weighted on a simple categorical dataset.
+ */
+BOOST_AUTO_TEST_CASE(CategoricalBuildTestWithWeight)
+{
+  arma::mat d;
+  arma::Row<size_t> l;
+  data::DatasetInfo di;
+  MockCategoricalData(d, l, di);
+  
   // Split into a training set and a test set.
   arma::mat trainingData = d.cols(0, 4999);
   arma::mat testData = d.cols(5000, 9999);
   arma::Row<size_t> trainingLabels = l.subvec(0, 4999);
   arma::Row<size_t> testLabels = l.subvec(5000, 9999);
 
+  arma::Row<double> weights = arma::ones<arma::Row<double>>(trainingLabels.n_elem);
+
   // Build the tree.
-  DecisionTree<> tree(trainingData, di, trainingLabels, 5, 10);
+  DecisionTree<> tree(trainingData, di, trainingLabels, 5, weights, 10);
```

Comment:
It would be fine in this test just to build another tree on unweighted data and ensure that the results are identical.